crates/flox-rust-sdk/src/models/environment/generations.rs

```diff
@@ -0,0 +1,328 @@
+    ) -> Result<(), GenerationsError> {
+        // keys should all be numbers (but)
+        let max = self
+            .metadata()?
+            .generations
+            .keys()
+            .cloned()
+            .max()
+            .unwrap_or_default();
+
+        self.register_generation(environment, *max + 1, generation_metadata, true)
+    }
+
+    /// Switch to a provided generation.
+    ///
+    /// Fails if the generation does not exist.
+    ///
+    /// This method will not perform any validation of the generation switched to.
+    /// If validation (e.g. proving that the environment builds) is required,
+    /// it should first be realized using [Self::realize_generation].
+    fn set_current_generation(&mut self, generation: usize) -> Result<(), GenerationsError> {
+        let mut metadata = self.metadata()?;
+        let realized = self.realize()?;
+
+        let generation_metadata = metadata.generations.contains_key(&generation.into());
+        if !generation_metadata {
+            return Err(GenerationsError::GenerationNotFound(generation));
+        }
+
+        metadata.current_gen = Some(generation.into());
+
+        write_metadata_file(metadata, realized.path())?;
+
+        realized.add(&[Path::new(".")]).unwrap();
+        realized
+            .commit(&format!("Set current generation to {}", generation))
+            .unwrap();
+        realized.push("origin", false).unwrap();
+
+        Ok(())
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum GenerationsError {
+    #[error("Generation {0} not found")]
+    GenerationNotFound(usize),
+
+    #[error("No generations found in environment")]
+    NoGenerations,
+}
+
+/// Reads the generations metadata file directly from the repository
+fn read_metadata(repo: &GitCommandProvider, ref_name: &str) -> Result<Metadata, GenerationsError> {
+    let metadata = {
+        let metadata_content = repo
+            .show(&format!("{}:{}", ref_name, GENERATIONS_METADATA_FILE))
+            .unwrap();
+        serde_json::from_str(&metadata_content.to_string_lossy()).unwrap()
+    };
+    Ok(metadata)
+}
+
+/// Serializes the generations metadata file to a path
+///
+/// The path is expected to be a realized generations repository.
+fn write_metadata_file(metadata: Metadata, realized_path: &Path) -> Result<(), GenerationsError> {
+    let metadata_content = serde_json::to_string(&metadata).unwrap();
+    let metadata_path = realized_path.join(GENERATIONS_METADATA_FILE);
+    fs::write(metadata_path, metadata_content).unwrap();
+    Ok(())
+}
+
+/// flox environment metadata for managed environments
+///
+/// Managed environments support rolling back to previous generations.
+/// Generations are defined immutable copy-on-write folders.
+/// Rollbacks and asssociated [GenerationMetadata] is tracked per environemnt
+/// in a metadata file at the root of the environment branch.
+#[derive(Serialize, Deserialize, Debug, Default)]
+#[serde(rename_all = "camelCase")]
+pub struct Metadata {
```

Comment:
Question: wasn't it a hard requirement that we be able to track the history of which environments were active at which point in time? It doesn't look like this metadata file tracks that. Is that supposed to be tracked by git history in the floxmeta branch?

Reply:
Yes we had that requirement, I'd like to circle back on that because **activating** the environment on two machines will create diverging environments which conflicts with pull/push.


@ghudgins @jennymahmoudi

Reply:
When the metadata.json file changes (and is committed) upon activation (i.e. to update the "lastActivated" field)

Reply:
I think we used to have a `lastActive` field that kept track of modifications to the environment, but as far as I can tell it wasn't changed when an environment was activated. I don't think we need to keep track of when an environment is activated, just when it is modified. It would be pretty undesirable if that caused environments to diverge

Reply:
> When the metadata.json file changes (and is committed) upon activation (i.e. to update the "lastActivated" field)

Hmmm ... the "lastActive" field was meant to indicate the last time that a generation was marked as the "current" generation. This has nothing at all to do with "flox activate". Is that the confusion?

Reply:
Alright, it's a naming confusion.
Setting as active != activate 

![image](https://github.com/flox/flox/assets/7040031/625e5886-673e-4f00-ac51-f022e6a1f07e)

Reply:
Ill add the lastActive back